### PR TITLE
feat(simd): int8 → int32 dot via VNNI / SDOT quad multiply-accumulate (#451 phase 3)

### PR DESCRIPTION
**Pull request by Ravenwater**
Phase 3 of epic #451 — the one the epic calls *"the case that matters"*. `dot<int32_t>` over 8-bit vectors now sums **four** products into each 32-bit lane with a single instruction, where phase 2's kernel sums two.

## The instruction is real

Verified as codegen, not inferred from values:

| `-march=` | emitted |
|---|---|
| `icelake-server` | **`vpdpbusd`** — native VNNI |
| `sapphirerapids` | **`vpdpbusd`** |
| `znver4` | **`vpdpbusd`** |
| `cascadelake` | `vpmaddwd` |
| `x86-64-v2` | `pmaddwd` |

The Cascade Lake row is worth knowing rather than glossing: it **has** AVX512-VNNI, but Highway's `HWY_AVX3_DL` target requires the full feature set (VBMI, VBMI2, GFNI, …), so VNNI hardware can still take the decomposed path. Correctness is identical either way; only the instruction count differs.

## Why 8 bits is the width that matters — headroom, not lanes

Halving the operand width quarters the product, so from the **same** 32-bit accumulator:

| pairing | max \|product\| | terms before the sum can overflow |
|---|---|---|
| `int16 × int16` *(phase 2)* | 2³⁰ | **3** worst case, ~36 random |
| `int8 × int8` | 2¹⁴ | **~131 000** |
| `uint8 × int8` | ~2¹⁵ | **~65 000** |
| `uint8 × uint8` | ~2¹⁶ | **~66 000** |

Four to five orders of magnitude. This is why the quantized-inference instructions are 8-bit and not 16-bit, and it is the concrete sense in which phase 2's kernel is for *small values in 16-bit storage* while this one is for *real vectors*.

Tested rather than asserted: 100 000 worst-case `int8` terms sum exactly without wrapping, 60 000 for the VNNI pairing, and the 16-bit contrast wraps at three. Past the headroom it still **wraps** — mod 2³², never a saturate, never a trap — and that is tested too.

## The signedness asymmetry is the hardware's

VNNI implements `unsigned × signed` — unsigned activations against signed weights, the shape of a quantized layer. Symmetric `int8 × int8` is native only from **AVX10.2**; before that Highway emulates it with two `dpbusd` calls plus a shift and subtract, roughly three times the work.

So MTL5 supports the three pairings the instruction supports, and **rejects `(int8, uint8)` at compile time** rather than silently reordering it. A dot product is symmetric, so swapping the arguments gets the native instruction — and the diagnostic says so:

```
error: unsupported 8-bit pairing -- the hardware op takes (i8,i8), (u8,i8) or
(u8,u8); for (i8,u8) swap the arguments, which a dot product permits
```

Exactly one diagnostic, not a cascade.

## The register tile does not need re-deriving

The epic's phase-3 sub-item is *"re-deriving the register tile for a 1-byte operand and a 4-byte accumulator"*. Investigated, and the answer is that there is nothing to re-derive:

`derive_blocking` sizes the tile from the **accumulator** — which `gemm_blocked<TC, TAB>` already passes it — and `int32` shares a lane width and a register file with `float`. So the int8 tile **is** the float tile, `mr × nr` identical at 128, 256 and 512 bits.

What a 1-byte operand actually constrains is the **K granularity**: the instruction consumes four k-values at once, so `kc` must be a multiple of 4 and the packed panels quad-interleaved (four k-values contiguous within each 32-bit word). `kc` is already a multiple of 4 in **all 93** cache configurations checked, so only the pack layout is new work for a future int8 GEMM.

Both facts are pinned as tests, so that work does not begin by hunting for a tile that needs no deriving.

## On the sequencing

The epic blocked this phase behind the blocking-model decision (#458). I checked that dependency rather than waiting on it: `derive_blocking` is a DAG — `(mr, nr) → kc → mc → nc` — and the tile depends only on the register file and FMA pipeline. Across 64 cache configurations × 3 types, `mr`/`nr` never moved while `kc`/`mc`/`nc` all did. #458 concerns `mc` and the thread partition, strictly downstream, so it cannot reach the tile. Recorded on [#458](https://github.com/stillwater-sc/mtl5/issues/458#issuecomment-5363246041) along with a second deterministic finding about `balanced_mc`.

## Changes

| File | What |
|---|---|
| `include/mtl/simd/batch.hpp` | `is_quad_widenable_v`, `quad_accumulator`/`QuadPair`, `quad_step`, `quad_dot_accumulate` on both backends |
| `include/mtl/simd/algorithm.hpp` | `reduce_dot_widen` generalized to two narrow types; 8-bit branch; the headroom table written down |
| `include/mtl/interface/dispatch_traits.hpp` | `SimdQuadVector` |
| `include/mtl/operation/dot.hpp` | `quad_int_dot` predicate; `dot`/`dot_real` select the kernel |
| `tests/unit/simd/test_quad_dot.cpp` | **new** — 10 cases, 258 assertions |
| `docs/algorithms/mixed-precision-kernels.md` | the 8-bit case, the signedness asymmetry, and what it means for an int8 GEMM |

`reduce_dot_widen` gained a **defaulted** third type parameter, so every phase 0/2 spelling remains source-compatible — verified by compiling them unchanged.

## Tests

| Compiler | Backend | Result |
|---|---|---|
| gcc 13.3 | scalar fallback | **175/175** |
| gcc 13.3 | Highway 1.4, `-march=native` | **175/175** |
| clang | scalar fallback | **175/175** |
| clang | Highway 1.4, `-march=native` | **175/175** |

Plus UBSan (`-fsanitize=undefined,integer`) over all the integer kernels.

25 lengths straddling the quad step on every backend; all three pairings against a closed form; partition invariance at all 138 cut points of a prime length, including splits landing mid-quad; the headroom claims; wrap-not-saturate past it; dispatch asserted directly including the three fallback cases.

## What this does not do

- No int8 **GEMM** — the tile is settled and the pack layout is identified, but building it is separate work
- No benchmarks — phase 4, deliberately last

## Test plan
- [ ] Tier 1 CI passes
- [ ] CodeRabbit review addressed
- [ ] Merge

Relates to #451

Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added optimized 8-bit integer dot products and GEMM support for supported signed and unsigned operand combinations.
  * Added quad-interleaved processing with 32-bit accumulation, including support for reversed operand pairings.
  * Added validation for compatible dense vector types and operand combinations.
* **Documentation**
  * Documented supported hardware instructions, accumulation behavior, overflow limits, and packing requirements.
* **Tests**
  * Added comprehensive coverage for accuracy, edge cases, fallback behavior, and GEMM packing constraints.

### CODE REVIEW

File: include/mtl/operation/dot.hpp [hunk 1]
```diff
@@ -58,6 +73,12 @@ auto dot(const V1& v1, const V2& v2) {
                       std::is_same_v<typename V2::value_type, float> &&
                       std::is_same_v<Accumulator, double> && std::is_same_v<Result, double>) {
             return simd::reduce_dot_widen<double, float>(v1.data(), v2.data(), v1.size());
+        } else if constexpr (quad_int_dot<Accumulator, Result, V1, V2>) {
+            // Quad-widening 8-bit fast path (#451 phase 3): VNNI / SDOT. conj is
+            // the identity on the integers, so this is the Hermitian product.
+            return simd::reduce_dot_widen<Accumulator, typename V1::value_type,
+                                          typename V2::value_type>(
+                v1.data(), v2.data(), v1.size());
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/mtl/operation/dot.hpp:105]: You found a real mismatch, and it was mine: the docs said `(int8, uint8)` is *"rejected at compile time"* without saying **where**. That is true of the kernel, `simd::reduce_dot_widen`, and false of `dot`/`dot_real`, which were quietly falling through to the generic accumulator loop. Fixed in aca03fd — but not the way you suggested, so let me give the reasoning rather than just diverge.

**Why not reject at the public API.** `dot` is generic over any `Vector` — it serves `dense_vector<posit>`, complex, custom number types. Refusing to compute a **well-defined dot product** because the hardware lacks a *fused* instruction would break generic code for a performance reason, and would be the only case where `dot` declines a mathematically valid call. The compile-time diagnostic belongs to the primitive that mirrors the instruction; it does not belong to the operation that means "inner product".

**Why the fallback was nonetheless unacceptable.** You are right that it could not stay as it was. Dropping to the generic loop costs roughly an order of magnitude against VNNI, and nothing tells the caller that one argument swap would have fixed it — precisely the silent detour the design set out to avoid. "Correct but 10× slow, and unremarked" is the worst of the options.

**Neither is necessary, because a dot product is symmetric.** `dot(i8, u8)` now dispatches to the native `(u8, i8)` instruction with the operands swapped:

```cpp
} else if constexpr (quad_int_dot_swapped<Accumulator, Result, V1, V2>) {
    return simd::reduce_dot_widen<Accumulator, typename V2::value_type,
                                  typename V1::value_type>(
        v2.data(), v1.data(), v1.size());
}
```

Exact, not approximate — identical products, identical sum — and `conj` is the identity on the integers, so it holds for the Hermitian `dot` as well as `dot_real`. Every caller now gets the native instruction regardless of argument order, and nothing that used to compile stops compiling.

The predicate is deliberately narrow: it fires only when the direct pairing is absent *and* the reversed one exists, so a pairing the hardware already has is never re-routed. Asserted both ways:

```cpp
STATIC_REQUIRE(mtl::quad_int_dot_swapped<int32_t, int32_t, i8v, u8v>);
STATIC_REQUIRE_FALSE(mtl::quad_int_dot_swapped<int32_t, int32_t, u8v, i8v>);
STATIC_REQUIRE_FALSE(mtl::quad_int_dot_swapped<int32_t, int32_t, i8v, i8v>);
```

Taking your four sites in turn:

- **`dot.hpp` L76-81 and L169-172** — both handled, by swap rather than rejection.
- **`test_quad_dot.cpp` L197-203** — the expectation that this pairing uses the generic loop is gone, replaced by the swap assertions above plus a value test showing `dot(a,b) == dot(b,a) == ` a 64-bit exact reference.
- **`mixed-precision-kernels.md` L282-290** — rewritten. The original had collapsed two levels into one claim; it now states the kernel's compile-time rejection and the public operations' swap separately, since that distinction is the whole point.

Full suite 175/175 on gcc and clang, each against both the Highway and scalar-fallback backends.
